File: npm/src/http.ts
import { createReadStream, statSync } from 'node:fs';
import { basename } from 'node:path';
import type { Readable } from 'node:stream';
import { BubleAPIError, BubleTimeoutError } from './errors.js';
import type { ApiErrorBody, BubleClientOptions, FetchLike, RequestOptions } from './types/common.js';
import type { UploadFileInput, UploadFileOptions } from './types/files.js';

type QueryValue = string | number | boolean | undefined | null;
type Query = Record<string, QueryValue>;

const DEFAULT_BASE_URL = 'https://buble.ai';
const DEFAULT_TIMEOUT = 60_000;

const FORBIDDEN_GENERATION_FIELDS = new Set([
  'input',
  'options',
  'scene',
  'sub_mode_id',
  'provider',
  'mediaType',
  'media_type',
  'images',
  'image_input',
  'video_input',
  'audio_input'
]);

function normalizeBaseURL(baseURL?: string) {
  return (baseURL || DEFAULT_BASE_URL).replace(/\/+$/, '');
}

function mergeHeaders(...items: Array<HeadersInit | undefined>) {
  const headers = new Headers();
  for (const item of items) {
    if (!item) continue;
    new Headers(item).forEach((value, key) => headers.set(key, value));
  }
  return headers;
}

function appendQuery(url: URL, query?: Query) {
  if (!query) return;
  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === null) continue;
    url.searchParams.set(key, String(value));
  }
}

async function parseResponse(response: Response) {
  const contentType = response.headers.get('content-type') || '';
  if (response.status === 204) return undefined;
  if (contentType.includes('application/json')) return response.json();
  return response.text();
}

function createTimeoutSignal(timeout: number, signal?: AbortSignal) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeout);

  if (signal) {
    if (signal.aborted) controller.abort(signal.reason);
    signal.addEventListener('abort', () => controller.abort(signal.reason), {
      once: true
    });
  }

  return {
    signal: controller.signal,
    clear: () => clearTimeout(timer)
  };
}

function assertNoForbiddenGenerationFields(body: Record<string, unknown>) {
  const forbidden = Object.keys(body).find((key) => FORBIDDEN_GENERATION_FIELDS.has(key));
  if (forbidden) {
    throw new BubleAPIError({
      status: 400,
      code: 'unsupported_field',
      message: `Field "${forbidden}" is an internal Buble field and is not supported by the public generation API.`,
      details: { field: forbidden }
    });
  }
}

function inferContentType(filename?: string) {
  const ext = filename?.split('.').pop()?.toLowerCase();
  if (!ext) return 'application/octet-stream';
  if (ext === 'png') return 'image/png';
  if (ext === 'jpg' || ext === 'jpeg') return 'image/jpeg';
  if (ext === 'webp') return 'image/webp';
  if (ext === 'gif') return 'image/gif';
  if (ext === 'mp4') return 'video/mp4';
  if (ext === 'mov') return 'video/quicktime';
  if (ext === 'webm') return 'video/webm';
  if (ext === 'mp3') return 'audio/mpeg';
  if (ext === 'wav') return 'audio/wav';
  return 'application/octet-stream';
}

function isBlob(value: unknown): value is Blob {
  return typeof Blob !== 'undefined' && value instanceof Blob;
}

function isReadable(value: unknown): value is NodeJS.ReadableStream {
  return Boolean(value && typeof (value as any).pipe === 'function');
}

function toUint8Array(value: string) {
  return new TextEncoder().encode(value);
}

async function* fileToIterable(file: UploadFileInput): AsyncIterable<Uint8Array> {
  if (typeof file === 'string') {
    yield* createReadStream(file) as unknown as AsyncIterable<Uint8Array>;
    return;
  }

  if (file instanceof ArrayBuffer) {
    yield new Uint8Array(file);
    return;
  }

  if (file instanceof Uint8Array) {
    yield file;
    return;
  }

  if (isBlob(file)) {
    for await (const chunk of file.stream() as unknown as AsyncIterable<Uint8Array>) {
      yield chunk;
    }
    return;
  }

  if (isReadable(file)) {
    for await (const chunk of file as Readable) {
      yield typeof chunk === 'string' ? toUint8Array(chunk) : new Uint8Array(chunk);
    }
    return;
  }

  throw new TypeError('Unsupported upload file input.');
}

function fileMetadata(file: UploadFileInput, options: UploadFileOptions) {
  if (typeof file === 'string') {
    const stat = statSync(file);
    const filename = options.filename || basename(file);
    return {
      filename,
      contentType: options.contentType || inferContentType(filename),
      size: stat.size
    };
  }

  if (isBlob(file)) {
    const filename = options.filename || 'file';
    return {
      filename,
      contentType: options.contentType || file.type || inferContentType(filename),
      size: file.size
    };
  }

  if (file instanceof ArrayBuffer || file instanceof Uint8Array) {
    const filename = options.filename || 'file';
    return {
      filename,
      contentType: options.contentType || inferContentType(filename),
      size: file.byteLength
    };
  }

  return {
    filename: options.filename || 'file',
    contentType: options.contentType || inferContentType(options.filename),
    size: undefined
  };
}

function multipartBody({
  file,
  fields,
  filename,
  contentType
}: {
  file: UploadFileInput;
  fields: Record<string, string>;
  filename: string;
  contentType: string;
}) {
  const boundary = `buble-sdk-${crypto.randomUUID()}`;

  async function* body() {
    for (const [name, value] of Object.entries(fields)) {
      yield toUint8Array(`--${boundary}\r\n`);
      yield toUint8Array(`Content-Disposition: form-data; name="${name}"\r\n\r\n`);
      yield toUint8Array(`${value}\r\n`);
    }

    yield toUint8Array(`--${boundary}\r\n`);
    yield toUint8Array(
      `Content-Disposition: form-data; name="file"; filename="${filename.replace(/"/g, '\\"')}"\r\n`
    );
    yield toUint8Array(`Content-Type: ${contentType}\r\n\r\n`);
    yield* fileToIterable(file);
    yield toUint8Array(`\r\n--${boundary}--\r\n`);
  }

  return {
    boundary,
    body: body() as unknown as BodyInit
  };
}

export class BubleHTTPClient {
  readonly apiKey: string;
  readonly baseURL: string;
  readonly timeout: number;
  readonly fetch: FetchLike;
  readonly headers?: HeadersInit;

  constructor(options: BubleClientOptions = {}) {
    const apiKey = options.apiKey || process.env.BUBLE_API_KEY;
    if (!apiKey) {
      throw new Error('Missing Buble API key. Pass apiKey or set BUBLE_API_KEY.');
    }

    if (!globalThis.fetch && !options.fetch) {
      throw new Error('Buble SDK requires fetch. Use Node.js 18+ or pass a custom fetch implementation.');
    }

    this.apiKey = apiKey;
    this.baseURL = normalizeBaseURL(options.baseURL || process.env.BUBLE_BASE_URL);
    this.timeout = options.timeout ?? DEFAULT_TIMEOUT;
    this.fetch = options.fetch || globalThis.fetch.bind(globalThis);
    this.headers = options.headers;
  }

  url(path: string, query?: Query) {
    const url = new URL(`${this.baseURL}${path.startsWith('/') ? path : `/${path}`}`);
    appendQuery(url, query);
    return url;
  }

  async request<T>(
    method: string,
    path: string,
    {
      query,
      body,
      headers,
      signal,
      timeout,
      raw
    }: RequestOptions & {
      query?: Query;
      body?: unknown;
      raw?: boolean;
    } = {}
  ): Promise<T> {
    const timeoutMs = timeout ?? this.timeout;
    const timeoutSignal = createTimeoutSignal(timeoutMs, signal);
    const requestHeaders = mergeHeaders(
      {
        Authorization: `Bearer ${this.apiKey}`,
        Accept: 'application/json'
      },
      this.headers,
      headers
    );

    let requestBody: BodyInit | undefined;
    if (body !== undefined) {
      requestHeaders.set('Content-Type', 'application/json');
      requestBody = JSON.stringify(body);
    }

    try {
      const response = await this.fetch(this.url(path, query), {
        method,
        headers: requestHeaders,
        body: requestBody,
        signal: timeoutSignal.signal
      });

      if (!response.ok) {
        const parsed = (await parseResponse(response)) as ApiErrorBody | string | undefined;
        const apiError = typeof parsed === 'object' ? parsed?.error : undefined;
        throw new BubleAPIError({
          status: response.status,
          code: apiError?.code,
          message: apiError?.message || response.statusText || 'Buble API request failed.',
          details: apiError?.details,
          response
        });
      }

      return (raw ? response : await parseResponse(response)) as T;
    } catch (error) {
      if (error instanceof BubleAPIError) throw error;
      if ((error as any)?.name === 'AbortError') {
        throw new BubleTimeoutError(`Buble API request timed out after ${timeoutMs}ms.`, timeoutMs);
      }
      throw error;
    } finally {
      timeoutSignal.clear();
    }
  }

  get<T>(path: string, query?: Query, options?: RequestOptions) {
    return this.request<T>('GET', path, { ...options, query });
  }

  post<T>(path: string, body?: unknown, options?: RequestOptions) {
    return this.request<T>('POST', path, { ...options, body });
  }

  async upload<T>(file: UploadFileInput, options: UploadFileOptions = {}): Promise<T> {
    const { filename, contentType } = fileMetadata(file, options);
    const fields: Record<string, string> = {};
    const fileType = options.file_type || options.fileType;

    if (fileType) fields.file_type = fileType;
    if (options.model) fields.model = options.model;
    if (options.mode) fields.mode = options.mode;

    const multipart = multipartBody({ file, fields, filename, contentType });
    const timeoutMs = options.timeout ?? this.timeout;
    const timeoutSignal = createTimeoutSignal(timeoutMs, options.signal);
    const headers = mergeHeaders(
      {
        Authorization: `Bearer ${this.apiKey}`,
        Accept: 'application/json',
        'Content-Type': `multipart/form-data; boundary=${multipart.boundary}`
      },
      this.headers
    );

    try {
      const response = await this.fetch(this.url('/api/v1/files'), {
        method: 'POST',
        headers,
        body: multipart.body,
        signal: timeoutSignal.signal,
        duplex: 'half'
      } as RequestInit & { duplex: 'half' });

      if (!response.ok) {
        const parsed = (await parseResponse(response)) as ApiErrorBody | string | undefined;
        const apiError = typeof parsed === 'object' ? parsed?.error : undefined;
        throw new BubleAPIError({
          status: response.status,
          code: apiError?.code,
          message: apiError?.message || response.statusText || 'Buble file upload failed.',
          details: apiError?.details,
          response
        });
      }

      return (await parseResponse(response)) as T;
    } catch (error) {
      if (error instanceof BubleAPIError) throw error;
      if ((error as any)?.name === 'AbortError') {
        throw new BubleTimeoutError(`Buble file upload timed out after ${timeoutMs}ms.`, timeoutMs);
      }
      throw error;
    } finally {
      timeoutSignal.clear();
    }
  }

  assertPublicGenerationBody(body: Record<string, unknown>) {
    assertNoForbiddenGenerationFields(body);
  }
}

File: python/src/buble_ai/_http.py
from __future__ import annotations

import os
from typing import Any, Dict, Iterator, Optional

import httpx

from ._errors import BubleAPIError, BubleTimeoutError

DEFAULT_BASE_URL = "https://buble.ai"
DEFAULT_TIMEOUT = 60.0


def _normalize_base_url(base_url: Optional[str]) -> str:
    return (base_url or DEFAULT_BASE_URL).rstrip("/")


def _api_error_from_response(response: httpx.Response) -> BubleAPIError:
    code = None
    message = response.reason_phrase or "Buble API request failed."
    details = None

    try:
      body = response.json()
    except ValueError:
      body = None

    if isinstance(body, dict) and isinstance(body.get("error"), dict):
        error = body["error"]
        code = error.get("code")
        message = error.get("message") or message
        details = error.get("details")
    elif response.text:
        message = response.text

    return BubleAPIError(
        message,
        status_code=response.status_code,
        code=code,
        details=details,
        response=response,
    )


class _BaseHTTP:
    def __init__(
        self,
        *,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        headers: Optional[Dict[str, str]] = None,
    ) -> None:
        resolved_api_key = api_key or os.environ.get("BUBLE_API_KEY")
        if not resolved_api_key:
            raise ValueError("Missing Buble API key. Pass api_key or set BUBLE_API_KEY.")

        self.api_key = resolved_api_key
        self.base_url = _normalize_base_url(base_url or os.environ.get("BUBLE_BASE_URL"))
        self.timeout = timeout
        self.headers = headers or {}

    def url(self, path: str) -> str:
        return f"{self.base_url}{path if path.startswith('/') else '/' + path}"

    def request_headers(self, extra: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
            **self.headers,
        }
        if extra:
            headers.update(extra)
        return headers


class BubleHTTPClient(_BaseHTTP):
    def __init__(
        self,
        *,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        headers: Optional[Dict[str, str]] = None,
        client: Optional[httpx.Client] = None,
    ) -> None:
        super().__init__(api_key=api_key, base_url=base_url, timeout=timeout, headers=headers)
        self.client = client or httpx.Client(timeout=timeout)
        self._owns_client = client is None

    def close(self) -> None:
        if self._owns_client:
            self.client.close()

    def request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Dict[str, Any]] = None,
        json: Any = None,
        data: Any = None,
        files: Any = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> Any:
        try:
            response = self.client.request(
                method,
                self.url(path),
                params=params,
                json=json,
                data=data,
                files=files,
                headers=self.request_headers(headers),
                timeout=timeout or self.timeout,
            )
        except httpx.TimeoutException as exc:
            raise BubleTimeoutError(
                f"Buble API request timed out after {timeout or self.timeout}s.",
                timeout=timeout or self.timeout,
            ) from exc

        if response.status_code >= 400:
            raise _api_error_from_response(response)

        if response.status_code == 204:
            return None
        content_type = response.headers.get("content-type", "")
        if "application/json" in content_type:
            return response.json()
        return response.text

    def stream(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> Iterator[str]:
        try:
            with self.client.stream(
                method,
                self.url(path),
                json=json,
                headers=self.request_headers({"Accept": "text/event-stream", **(headers or {})}),
                timeout=timeout or self.timeout,
            ) as response:
                if response.status_code >= 400:
                    response.read()
                    raise _api_error_from_response(response)
                for line in response.iter_lines():
                    yield line
        except httpx.TimeoutException as exc:
            raise BubleTimeoutError(
                f"Buble API stream timed out after {timeout or self.timeout}s.",
                timeout=timeout or self.timeout,
            ) from exc


class AsyncBubleHTTPClient(_BaseHTTP):
    def __init__(
        self,
        *,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        headers: Optional[Dict[str, str]] = None,
        client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        super().__init__(api_key=api_key, base_url=base_url, timeout=timeout, headers=headers)
        self.client = client or httpx.AsyncClient(timeout=timeout)
        self._owns_client = client is None

    async def close(self) -> None:
        if self._owns_client:
            await self.client.aclose()

    async def request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Dict[str, Any]] = None,
        json: Any = None,
        data: Any = None,
        files: Any = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> Any:
        try:
            response = await self.client.request(
                method,
                self.url(path),
                params=params,
                json=json,
                data=data,
                files=files,
                headers=self.request_headers(headers),
                timeout=timeout or self.timeout,
            )
        except httpx.TimeoutException as exc:
            raise BubleTimeoutError(
                f"Buble API request timed out after {timeout or self.timeout}s.",
                timeout=timeout or self.timeout,
            ) from exc

        if response.status_code >= 400:
            raise _api_error_from_response(response)

        if response.status_code == 204:
            return None
        content_type = response.headers.get("content-type", "")
        if "application/json" in content_type:
            return response.json()
        return response.text

    async def stream(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None,
    ):
        try:
            async with self.client.stream(
                method,
                self.url(path),
                json=json,
                headers=self.request_headers({"Accept": "text/event-stream", **(headers or {})}),
                timeout=timeout or self.timeout,
            ) as response:
                if response.status_code >= 400:
                    await response.aread()
                    raise _api_error_from_response(response)
                async for line in response.aiter_lines():
                    yield line
        except httpx.TimeoutException as exc:
            raise BubleTimeoutError(
                f"Buble API stream timed out after {timeout or self.timeout}s.",
                timeout=timeout or self.timeout,
            ) from exc

